<a href="https://colab.research.google.com/github/slisovych/machine_learning_hometasks/blob/main/HW_2_1_%D0%A2%D0%B5%D1%81%D1%82%D1%83%D0%B2%D0%B0%D0%BD%D0%BD%D1%8F_%D1%81%D1%82%D0%B0%D1%82_%D0%B3%D1%96%D0%BF%D0%BE%D1%82%D0%B5%D0%B7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [2]:
import numpy as np
from statsmodels.stats.weightstats import ztest

In [3]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [11]:
null_mean = 170
alpha = 0.05
sample_size = len(heights)
sample_mean = np.mean(heights)
sample_std = np.std(heights, ddof=1)

print ("Sample_size: {:.2f}".format(sample_size))
print ("Sample_mean: {:.2f}".format(sample_mean))
print ("Sample_std: {:.2f}".format(sample_std))

Sample_size: 33.00
Sample_mean: 170.70
Sample_std: 4.72


In [13]:
ztest_Score, p_value = ztest(heights, value = null_mean, alternative='two-sided')

In [14]:
ztest_Score, p_value

(0.8482235371587619, 0.3963134914218349)

In [15]:
if(p_value < alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [16]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [17]:
np.mean(group_a_scores), np.mean(group_b_scores)

(77.08125, 74.73875000000001)

In [24]:
ztest_Score, p_value = ztest(group_a_scores, group_b_scores, alternative='larger')

In [25]:
ztest_Score, p_value

(2.357354864751502, 0.009202824276264005)

In [27]:
ztest_Score, p_value = ztest(group_a_scores, group_b_scores, alternative='two-sided')
ztest_Score, p_value

(2.357354864751502, 0.01840564855252801)

Оскільки p-value < 0.05, у нас є достатньо доказів, щоб відкинути нульову гіпотезу. Новий курс у групі А є ефективнішим.

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [28]:
from scipy import stats

In [29]:
group_a_scores_5 = group_a_scores[:5]
group_a_scores_5

[78.55, 72.25, 79.88, 75.0, 76.54]

In [30]:
stats.ttest_ind(group_a_scores_5, group_b_scores)

TtestResult(statistic=0.8167951030410814, pvalue=0.4185479477535733, df=43.0)

In [31]:
stats.ttest_ind(group_a_scores_5, group_b_scores, equal_var=False)

TtestResult(statistic=1.1226527301193456, pvalue=0.3010736990344312, df=6.544857383782639)

p-value для обох тестів значно більше 0.05:

- для тесту з рівними дисперсіями: p = 0.4185

- для тесту з нерівними дисперсіями: p = 0.3011


У жодному з випадків ми не можемо відхилити нульову гіпотезу. Це означає, що статистично значущої різниці між групами A та B немає.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [32]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [33]:
difference = np.array(after_training) - np.array(before_training)
difference

array([ 4.65,  3.03,  5.9 ,  2.47,  8.44,  4.26,  1.29,  0.86,  3.97,
        2.55,  4.42,  3.94,  2.85,  1.3 , -0.03,  2.11,  4.71,  3.42,
        0.51,  3.35,  3.77,  1.23,  3.31,  3.11,  0.72,  3.72,  4.12,
        5.17,  5.11,  0.24])

In [34]:
mean_difference = np.mean(difference)
std_difference = np.std(difference, ddof=1)
size_difference = len(difference)
print ("Sample_mean: {:.2f}".format(mean_difference))
print ("Sample_std: {:.2f}".format(std_difference))
print ("Sample_size: {:.2f}".format(size_difference))


Sample_mean: 3.15
Sample_std: 1.89
Sample_size: 30.00


In [39]:
ztest_Score, p_value = ztest(difference, value = 0, alternative='larger')
ztest_Score, p_value

(9.13890009623916, 3.1545359919175505e-20)

In [40]:
if(p_value < alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Відхилити Н0.


Оскільки p-значення <0.05, ми відхиляємо нульову гіпотезу.
текст курсивом
Є статистично значущий доказ, що тренінг позитивно вплинув на продуктивність працівників.